crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -126,10 +126,20 @@ pub(super) fn union_or_intersection_elements_ordering<'db>(
 
         (Type::SubclassOf(_), _) => Ordering::Less,
         (_, Type::SubclassOf(_)) => Ordering::Greater,
-        (Type::Instance(left), Type::Instance(right)) => left.class().cmp(&right.class()),
 
-        (Type::Instance(_), _) => Ordering::Less,
-        (_, Type::Instance(_)) => Ordering::Greater,
+        (Type::NominalInstance(left), Type::NominalInstance(right)) => {
+            left.class().cmp(&right.class())
+        }
+        (Type::NominalInstance(_), _) => Ordering::Less,
+        (_, Type::NominalInstance(_)) => Ordering::Greater,
+
+        (Type::ProtocolInstance(left_proto), Type::ProtocolInstance(right_proto)) => {
+            debug_assert_eq!(*left, left_proto.normalized(db));
+            debug_assert_eq!(*right, right_proto.normalized(db));
```

Comment:
Is it generally invariant that we only order normalized types? Is there a more general way we should enforce/check that invariant?

Reply:
> Is it generally invariant that we only order normalized types?

Yes. It doesn't really make sense to order a non-normalized `UnionType`, for example, because it might contain intersections that are not ordered (or Tuples that contain unions that are not ordered, or tuples that contain unions that contain intersections that are not ordered... you get my drift :-).

> Is there a more general way we should enforce/check that invariant?

Yes, good point. I can update the various `debug_assert`s in this function.

Reply:
https://github.com/astral-sh/ruff/pull/17734